packages/cluster-ui/src/store/terminateQuery/terminateQuery.sagas.ts

```diff
@@ -11,49 +11,74 @@
 import { PayloadAction } from "@reduxjs/toolkit";
 import { all, call, put, takeEvery } from "redux-saga/effects";
 
-// TODO: (vlad) originaly this calls triggered popup message with error\success message.
-// need to come up with api to recreate this behavier with client specific popup components
-// import {terminateQueryAlertLocalSetting, terminateSessionAlertLocalSetting} from "src/redux/alerts";
-
 import { terminateQuery, terminateSession } from "src/api/terminateQueryApi";
 import { actions as sessionsActions } from "src/store/sessions";
 import { actions as terminateQueryActions } from "./terminateQuery.reducer";
 
 import { cockroach } from "@cockroachlabs/crdb-protobuf-client";
+import { XOR } from "src/util/xor";
+
 const CancelSessionRequest = cockroach.server.serverpb.CancelSessionRequest;
 const CancelQueryRequest = cockroach.server.serverpb.CancelQueryRequest;
-export type ICancelSessionRequest = cockroach.server.serverpb.CancelSessionRequest;
-export type ICancelQueryRequest = cockroach.server.serverpb.CancelQueryRequest;
+export type ICancelSessionRequest = cockroach.server.serverpb.ICancelSessionRequest;
+export type ICancelQueryRequest = cockroach.server.serverpb.ICancelQueryRequest;
+interface CancelSessionWithCallbacks {
+  req: ICancelSessionRequest;
+  onCancelSessionSuccess?: () => void;
+  onCancelSessionError?: () => void;
+}
+
+export type CancelSessionPayload = XOR<
+  CancelSessionWithCallbacks,
+  ICancelSessionRequest
+>;
 
 export function* terminateSessionSaga(
-  action: PayloadAction<ICancelSessionRequest>,
+  action: PayloadAction<CancelSessionPayload>,
 ) {
-  const terminateSessionRequest = new CancelSessionRequest(action.payload);
+  const req = action.payload.req || action.payload;
+  const { onCancelSessionSuccess, onCancelSessionError } = action.payload;
+  const terminateSessionRequest = new CancelSessionRequest(req);
+
   try {
     yield call(terminateSession, terminateSessionRequest);
     yield put(terminateQueryActions.terminateSessionCompleted());
     yield put(sessionsActions.invalidated());
     yield put(sessionsActions.refresh());
+    onCancelSessionSuccess && onCancelSessionSuccess();
```

Comment:
If we just have a single notification dispatch can we avoid having to do all the null checks?

Reply:
Not really, still need null check to support api for dbconsole env. Will be not necessary if/when we move redux layer to CC.